### ISSUE TITLE
Mi Band 2: Display items

### ISSUE THREAD

**Issue by ecneb85** [Freeyourgadget/Gadgetbridge#784]
#### Before opening an issue please confirm the following:
- [x] I have read the [wiki](https://github.com/Freeyourgadget/Gadgetbridge/wiki), and I didn't find a solution to my problem / an answer to my question.
- [x] I have searched the [issues](https://github.com/Freeyourgadget/Gadgetbridge/issues), and I didn't find a solution to my problem / an answer to my question.
- [ ] If you upload an image or other content, please make sure you have read and understood the [github policies and terms of services](https://help.github.com/articles/github-terms-of-service/#1-responsibility-for-user-generated-content)

#### Your issue is:
Mi band settings ->  Display items
Cannot set up Mi Band 2 to show Distance and Calories also on display.
log:
07:31:33.494 [main] DEBUG n.f.g.s.DeviceCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbridge.devices.action.send_configuration
07:31:33.495 [main] INFO  n.f.g.s.d.m.MiBand2Support - Setting display items to [calories, steps, heart_rate, distance, battery]
07:31:33.497 [main] DEBUG n.f.g.s.b.BtLEQueue - about to add: 7:31:33 AM: Transaction task: Sending configuration for option: mi2_display_items with 1 actions
07:31:33.499 [Gadgetbridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - About to run action: August 31, 07:31: WriteAction on characteristic: 00000003-0000-3512-2118-0009af100700
07:31:33.500 [Gadgetbridge GATT Dispatcher] DEBUG n.f.g.s.b.a.WriteAction - writing to characteristic: 00000003-0000-3512-2118-0009af100700: 0x a 0x3f 0x 0 0x 0 0x 1 0x 2 0x 3 0x 4 0x 5
07:31:33.503 [Binder:6730_1] DEBUG n.f.g.s.b.BtLEQueue - characteristic write: 00000003-0000-3512-2118-0009af100700 (success)
07:31:33.716 [Binder:6730_2] DEBUG n.f.g.s.b.BtLEQueue - characteristic changed: 00000003-0000-3512-2118-0009af100700 value:  0x10 0xa 0x1
07:31:33.717 [Binder:6730_2] INFO  n.f.g.s.d.m.MiBand2Support - Unhandled characteristic changed: 00000003-0000-3512-2118-0009af100700
07:31:33.717 [Binder:6730_2] INFO  n.f.g.s.d.m.MiBand2Support - RECEIVED DATA WITH LENGTH: 3

#### Your wearable device is:
Mi BAND 2
FW: 1.0.1.39
HW: V0.1.3.3

#### Your android version is:
7.0 (Samsung S6)
#### Your Gadgetbridge version is:
0.20.1

*New issues about already solved/documented topics could be closed without further comments. Same for too generic or incomplete reports.*

**Comment by svenburkard** [comment]
hi, i have exactly the same problem with a new xiaomi mi band 2.

**issue:**
i can change the items to display, except for distance and calories.

**wearable device:**
xiaomi mi band 2
fw: 1.0.1.53
hw: v0.1.3.3

**android version:**
7.1.2 (nexus 7)

**gadgetbridge version:**
0.21.1

**log (set to display calories, steps, heart_rate, distance, battery => calories and distance will not show up):**
13:35:13.761 [main] ERROR n.f.g.a.AbstractSettingsActivity - Unknown preference key: mi_vibration_count_generic_alarm_clock, unable to display value.
13:35:27.771 [main] DEBUG n.f.g.s.DeviceCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbridge.devices.action.send_configuration
13:35:27.773 [main] INFO  n.f.g.s.d.m.MiBand2Support - Setting display items to [calories, steps, heart_rate, distance, battery]
13:35:27.776 [main] DEBUG n.f.g.s.b.BtLEQueue - about to add: 1:35:27 nachm.: Transaction task: Sending configuration for option: mi2_display_items with 1 actions
13:35:27.779 [Gadgetbridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - About to run action: 20. September, 13:35: WriteAction on characteristic: 00000003-0000-3512-2118-0009af100700
13:35:27.785 [Gadgetbridge GATT Dispatcher] DEBUG n.f.g.s.b.a.WriteAction - writing to characteristic: 00000003-0000-3512-2118-0009af100700: 0x a 0x3f 0x 0 0x 0 0x 1 0x 2 0x 3 0x 4 0x 5
13:35:27.789 [Binder:11224_3] DEBUG n.f.g.s.b.BtLEQueue - characteristic write: 00000003-0000-3512-2118-0009af100700 (success)
13:35:28.381 [Binder:11224_2] DEBUG n.f.g.s.b.BtLEQueue - characteristic changed: 00000003-0000-3512-2118-0009af100700 value:  0x10 0xa 0x1
13:35:28.382 [Binder:11224_2] INFO  n.f.g.s.d.m.MiBand2Support - Unhandled characteristic changed: 00000003-0000-3512-2118-0009af100700
13:35:28.384 [Binder:11224_2] INFO  n.f.g.s.d.m.MiBand2Support - RECEIVED DATA WITH LENGTH: 3
13:35:28.384 [Binder:11224_2] WARN  n.f.g.s.d.m.MiBand2Support - DATA: 0x10
13:35:28.385 [Binder:11224_2] WARN  n.f.g.s.d.m.MiBand2Support - DATA: 0x a
13:35:28.387 [Binder:11224_2] WARN  n.f.g.s.d.m.MiBand2Support - DATA: 0x 1

**log (set to display only battery => works like expected):**
13:34:36.516 [main] DEBUG n.f.g.s.DeviceCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbridge.devices.action.send_configuration
13:34:36.517 [main] INFO  n.f.g.s.d.m.MiBand2Support - Setting display items to [battery]
13:34:36.518 [main] DEBUG n.f.g.s.b.BtLEQueue - about to add: 1:34:36 nachm.: Transaction task: Sending configuration for option: mi2_display_items with 1 actions
13:34:36.520 [Gadgetbridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - About to run action: 20. September, 13:34: WriteAction on characteristic: 00000003-0000-3512-2118-0009af100700
13:34:36.533 [Gadgetbridge GATT Dispatcher] DEBUG n.f.g.s.b.a.WriteAction - writing to characteristic: 00000003-0000-3512-2118-0009af100700: 0x a 0x21 0x 0 0x 0 0x 1 0x 2 0x 3 0x 4 0x 5
13:34:36.536 [Binder:11224_3] DEBUG n.f.g.s.b.BtLEQueue - characteristic write: 00000003-0000-3512-2118-0009af100700 (success)
13:34:37.153 [Binder:11224_2] DEBUG n.f.g.s.b.BtLEQueue - characteristic changed: 00000003-0000-3512-2118-0009af100700 value:  0x10 0xa 0x1
13:34:37.153 [Binder:11224_2] INFO  n.f.g.s.d.m.MiBand2Support - Unhandled characteristic changed: 00000003-0000-3512-2118-0009af100700
13:34:37.155 [Binder:11224_2] INFO  n.f.g.s.d.m.MiBand2Support - RECEIVED DATA WITH LENGTH: 3
13:34:37.156 [Binder:11224_2] WARN  n.f.g.s.d.m.MiBand2Support - DATA: 0x10
13:34:37.157 [Binder:11224_2] WARN  n.f.g.s.d.m.MiBand2Support - DATA: 0x a
13:34:37.157 [Binder:11224_2] WARN  n.f.g.s.d.m.MiBand2Support - DATA: 0x 1

**Comment by joserebelo** [comment]
I can't reproduce this, and the logs seem fine. The only difference I can see here is the HW version, both of you have V0.1.3.3, while my band is V0.1.3.2.

Could someone else with V0.1.3.3 reproduce this problem?

**Comment by Kokokokoka** [comment]
I've got HW 0.1.3.2 and FW 1.0.1.50 (as far as I can remember I've been using *.39 before with no luck and I was thinking that 1.50 could have those measurements). There is no distance and calories in my band too.

**Comment by joserebelo** [comment]
@Kokokokoka thanks for the confirmation, it was just a shot in the dark.

Would someone with this problem be able to capture Bluetooth HCI logs when using Mi Fit to change the display items?

**Comment by ecneb85** [comment]
First part: 
battery status OFF
all other items ON
Second part:
all items ON

MI Fit 3.0.5
Firmware: V1.0.1.53

[btsnoop_hci.txt](https://github.com/Freeyourgadget/Gadgetbridge/files/1320620/btsnoop_hci.txt)

**Comment by joserebelo** [comment]
@ecneb85 thanks. Unless I'm missing something though, I see nothing wrong that might be causing this :(

Could you by any chance get the same log, now from Gadgetbridge changing the items?

**Comment by ecneb85** [comment]
I can´t find in source why there are those wrong HEX string formattings in GB log files, I mean "0x a" it should be "0x0a" or "0xa".
And why there are INFO about unhandled characteric changes.
13:34:37.153 [Binder:11224_2] INFO n.f.g.s.d.m.MiBand2Support - Unhandled characteristic changed: 00000003-0000-3512-2118-0009af100700

**Comment by joserebelo** [comment]
The Unhandled characteristic changed is related to #715, it indicates that the characteristic was changed successfully. See [here](https://github.com/Freeyourgadget/Gadgetbridge/blob/6ec15551781046a389bf695f68b219a8ed6995db/app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/miband2/MiBand2Support.java#L887).

As for the formatting error, i'd say 0x0a would look better, but that's just a formatting problem caused by Loggin.formatBytes [here](https://github.com/Freeyourgadget/Gadgetbridge/blob/6ec15551781046a389bf695f68b219a8ed6995db/app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/btle/actions/WriteAction.java#L58).

**Comment by ecneb85** [comment]
I have tried again with GB and the setting of the displayed items works fine with the new Firmware.
Firmware: V1.0.1.53
GB 0.21.0

**Comment by joserebelo** [comment]
That's weird, because @svenburkard still has the problem in 1.0.1.53.

If someone is able to reproduce this problem, please try to capture Bluetooth HCI Logs while using Gadgetbridge to change the display items.

**Comment by ecneb85** [comment]
Now I see. It's works because I have already enabled these items with Mi Fit.
If I disable all items with Mi Fit then I have the same issue again.

**Comment by joserebelo** [comment]
Okay, I might be able to reproduce the problem then. I will install Mi Fit and try to figure out what is wrong.

**Comment by svenburkard** [comment]
i can't capture bluetooth hci logs right now, but maybe it's useful to know that i never installed mi fit and configured the device only with gadgetbridge.

**Comment by Kokokokoka** [comment]
Can confirm that I have never used the Mi Fit app too.

**Comment by svenburkard** [comment]
ok, i'm back around my laptop and captured some bluetooth hci logs.

i captured 3 cases and before starting to capture any of them i deactivated all settings, to start always without any display options activated.

[btsnoop_hci_-_only_distance_set.log.txt](https://github.com/Freeyourgadget/Gadgetbridge/files/1321359/btsnoop_hci_-_only_distance_set.log.txt)

[btsnoop_hci_-_only_calories_set.log.txt](https://github.com/Freeyourgadget/Gadgetbridge/files/1321358/btsnoop_hci_-_only_calories_set.log.txt)

[btsnoop_hci_-_all_set.log.txt](https://github.com/Freeyourgadget/Gadgetbridge/files/1321357/btsnoop_hci_-_all_set.log.txt)

**Comment by joserebelo** [comment]
I still can't reproduce this, but I think the problem isn't related to the changing of the display items.

The Mi Band 1 has the user info sent to the band, and that is not done for the Mi Band 2. Mi Fit is probably sending it correctly, and after using Mi Fit at least once the display item setting will work. The band probably needs the user's height / weight in order to calculate and display the distance and calories correctly.

@cpfeiffer any thoughts / ideas on this?

**Comment by Kokokokoka** [comment]
I was thinking about height and etc. Still I changed those settings in gadgetbridge, don't know whether it works or not though.

**Comment by joserebelo** [comment]
@Kokokokoka Gadgetbridge doesn't send that information to the Mi Band 2. I've tried looking into the Mi Fit pairing logs to figure out if it should but without success so far.

**Comment by ashimokawa** [comment]
@joserebelo 
I will look into how to send user info. Didn't know we don't do this.
Since the Bip is protocol compatible I should be able to test.

**Comment by ashimokawa** [comment]
@joserebelo 
I have implemented setting user info in master (it is disabled, because it only sends test data instead of taking it from preferences)

@ecneb85 
@svenburkard 
@cpfeiffer 
Dare to test?

**Comment by ashimokawa** [comment]
It is now enabled in master and takes information from settings.
Hope it works

**Comment by svenburkard** [comment]
thank you @ashimokawa, @joserebelo and all others involved.
i built the latest master (46b5051) and can confirm that it's working now with my mi band 2.

**Comment by ashimokawa** [comment]
Thanks, closing this ;)

**Comment by custoz** [comment]
Hi all,

I see this has been closed almost one year ago, however, I still have the same (actually similar) issue: I can change the items to display, including distance and calories, but actually these two do not appear on the mi band 2

wearable device:
xiaomi mi band 2
fw: 1.0.1.81
hw: v0.1.3.3

android version:
8.0.0 (Honor 9)

gadgetbridge version:
0.28.1

**Comment by MagicFab** [comment]
Works for me, although I have slightly different versions:

Xiaomi MiBand 2
F/W: 1.0.1.81
H/W: v0.1.3.**2**

Android version:
8.**1**.0 (Lineage OS 15.1)

Gadgetbridge version:
0.28.1

Consider reporting this as a new issue.

**Comment by custoz** [comment]
Thanks a lot for the immediate responses! 

Indeed, it was my fault and @cpfeiffer helped me in tracking my mistake. I actually did two installations in a row, but in the second one I forgot to specify height and weight (facepalm). It then required me to manually restart the app a couple of time to work. 

Again, thanks for helping me :)
